server/app/assets/javascripts/session.ts

```diff
@@ -0,0 +1,396 @@
+import {ToastController} from './toast'
+
+/**
+ * Represents session timeout data with timestamps for various timeout events.
+ * All timestamps are Unix timestamps in seconds.
+ */
+interface TimeoutData {
+  /** When to show the inactivity warning modal */
+  inactivityWarning: number
+  /** When to logout due to inactivity */
+  inactivityTimeout: number
+  /** When to show the total session length warning modal */
+  totalWarning: number
+  /** When to logout due to total session length */
+  totalTimeout: number
+  /** Server's current time when cookie was set - used for clock skew calculation */
+  currentTime: number
+}
+
+/**
+ * Types of warning modals that can be shown to the user.
+ */
+export enum WarningType {
+  /** Warning shown when user has been inactive */
+  INACTIVITY = 'inactivity',
+  /** Warning shown when total session length is approaching limit */
+  TOTAL_LENGTH = 'total_length',
+}
+
+/**
+ * Handles session timeout management including:
+ * - Reading timeout data from cookies
+ * - Showing warning modals before timeout
+ * - Handling session extension requests
+ * - Managing automatic logout
+ */
+export class SessionTimeoutHandler {
+  /** Name of cookie storing timeout data */
+  private static COOKIE_NAME = 'session_timeout_data'
```

Comment:
suggestion (non-blocking): name this var `TIMEOUT_COOKIE_NAME` to match the one in `SessionTimeoutFilter.java` and add a comment noting that they must be kept in sync. [Here's](https://github.com/civiform/civiform/blob/1f7644c8116708a88a176f3eb42d116a0a172694/server/app/assets/javascripts/admin_export_view.ts#L4) an example of where we do this elsewhere